### PR TITLE
Don't hardcode existing relations in TextAnalysisFormat

### PR DESCRIPTION
**Pull request by Duhemm**
This follows my work on making it easier to add new dependency kinds in sbt.

> The previous implementation of TextAnalysisFormat contained the list
> of all the existing relations that sbt knew of, and used this
> information to write to and read from the disk the persisted analyses.
> 
> In this knew implementation, TextAnalysisFormat gets from the
> Relations object what are the existing relations, and then persists
> them to disk.
> 
> The previous situation was not optimal since it meant that, in order
> to add a new dependency kind, one had to modify both the Relations
> and TextAnalysisFormat.
> 
> Using this new implementation, no change to TextAnalysisFormat is
> required whenever a new dependency kind is added.

Thanks @gkossakowski for the help !

/cc @xeno-by

### CODE REVIEW

**Comment by typesafe-tools** [comment]
Can one of the admins verify this patch?

**Comment by Duhemm** [comment]
I don't know what went wrong with the build : https://travis-ci.org/sbt/sbt/jobs/34338983#L3380
Could someone please restart it ?

Speaking of tests, I don't know what are the criteria to include unit tests in the `safeUnitTests` command. Should `persist/test` be considered a safeUnitTest  ?

File: compile/inc/src/main/scala/sbt/inc/Relations.scala [hunk 1]
```diff
@@ -175,9 +175,66 @@ trait Relations {
...
+      ("public inherited external dependencies", identity[String] _),
+      ("member reference internal dependencies", string2File),
+      ("member reference external dependencies", identity[String] _),
+      ("inheritance internal dependencies", string2File),
+      ("inheritance external dependencies", identity[String] _),
+      ("class names", identity[String] _),
+      ("used names", identity[String] _))
+  }
+  /**
+   * Reconstructs a Relations from a list of Relation
```
Comment (gkossakowski) [inline compile/inc/src/main/scala/sbt/inc/Relations.scala:210]: Could add a comment saying that the order of Relations matters and is defined by `existingRelations`?

Comment (Duhemm) [inline compile/inc/src/main/scala/sbt/inc/Relations.scala:210]: Done.

**Comment by gkossakowski** [comment]
Thanks. More comments to follow today. My iPhone got bricked and my life is in chaos now... :rage1:

File: compile/inc/src/main/scala/sbt/inc/Relations.scala [hunk 2]
```diff
@@ -403,6 +461,23 @@ private class MRelationsDefaultImpl(srcProd: Relation[File, File], binaryDep: Re
...
 
+  def allRelations = {
+    val rels = List(
+      srcProd,
+      binaryDep,
+      direct.internal,
+      direct.external,
+      publicInherited.internal,
+      publicInherited.external,
+      Relations.emptySourceDependencies.internal, // Default implementation doesn't provide memberRef source deps
```
Comment (gkossakowski) [inline compile/inc/src/main/scala/sbt/inc/Relations.scala:472]: I don't understand this comment. Could you explain?

Comment (gkossakowski) [inline compile/inc/src/main/scala/sbt/inc/Relations.scala:472]: Scratch that. Now I get it.

File: compile/inc/src/main/scala/sbt/inc/Relations.scala [hunk 3]
```diff
@@ -175,9 +175,67 @@ trait Relations {
...
+  private[inc] def allRelations: List[(String, Relation[File, _])]
 }
 
 object Relations {
+
+  /**
+   * Represents all the relations that sbt knows of along with a way to recreate each
+   * of their elements from their string representation.
+   */
+  val existingRelations = {
```
Comment (gkossakowski) [inline compile/inc/src/main/scala/sbt/inc/Relations.scala]: Let's mark this `privete[inc]` too so we have freedom to modify it in the future.

File: compile/inc/src/main/scala/sbt/inc/Relations.scala [hunk 4]
```diff
@@ -175,9 +175,67 @@ trait Relations {
...
+      ("class names", identity[String] _),
+      ("used names", identity[String] _))
+  }
+  /**
+   * Reconstructs a Relations from a list of Relation
+   * The order in which the relations are read matters and is defined by `existingRelations`.
+   */
+  def construct(nameHashing: Boolean, relations: List[Relation[_, _]]) =
+    relations match {
+      case p :: bin :: di :: de :: pii :: pie :: mri :: mre :: ii :: ie :: cn :: un :: Nil =>
```
Comment (gkossakowski) [inline compile/inc/src/main/scala/sbt/inc/Relations.scala:215]: This relies on matching the order defined in `existingRelations` so it's a bit fragile. I don't have any good idea how to address that, though. @Duhemm, if an improvement comes to your mind please submit a subsequent PR. We won't be blocking this one.

**Comment by gkossakowski** [comment]
Overall looks good. Please address my one more comment, squash commits and this will be good to go. :+1:

**Comment by Duhemm** [comment]
Great, thanks ! Should I squash all commits in one or keep the original 3 commits ?

**Comment by gkossakowski** [comment]
No, let's keep separate commits for separate things. I just meant the last one where you addressed my feedback. I see you squashed that already so let's wait for tests and we can merge.

**Comment by gkossakowski** [comment]
Forgot to say: LGTM. :-)

**Pull request merged**